src/test/java/teammates/common/datatransfer/questions/FeedbackContributionQuestionDetailsTest.java

```diff
@@ -21,4 +37,568 @@ public void testIsFeedbackParticipantCommentsOnResponsesAllowed_shouldReturnFals
+            details.setAnswer(answer);
+            responses.add(details);
+            expectedResponsesValidationResults.add(FeedbackContributionQuestionDetails.CONTRIB_ERROR_INVALID_OPTION);
+        }
+        assertEquals(expectedResponsesValidationResults,
+                feedbackContributionQuestionDetails.validateResponsesDetails(responses, 0));
+
+    }
+
+    @Test
+    public void testValidateGiverRecipientVisibility() {
+        FeedbackContributionQuestionDetails details = new FeedbackContributionQuestionDetails();
+        FeedbackQuestionAttributes feedbackQuestionAttributes = FeedbackQuestionAttributes.builder()
+                .withCourseId("course")
+                .withFeedbackSessionName("session")
+                .withGiverType(FeedbackParticipantType.STUDENTS)
+                .withRecipientType(FeedbackParticipantType.OWN_TEAM_MEMBERS_INCLUDING_SELF)
+                .withQuestionNumber(1)
+                .withNumberOfEntitiesToGiveFeedbackTo(Const.MAX_POSSIBLE_RECIPIENTS)
+                .withShowResponsesTo(Arrays.asList(FeedbackParticipantType.RECEIVER,
+                        FeedbackParticipantType.RECEIVER_TEAM_MEMBERS,
+                        FeedbackParticipantType.OWN_TEAM_MEMBERS,
+                        FeedbackParticipantType.INSTRUCTORS))
+                .withShowGiverNameTo(new ArrayList<>())
+                .withShowRecipientNameTo(new ArrayList<>())
+                .withQuestionDescription("description")
+                .build();
+
+        ______TS("success: valid giver recipient visibility");
+        assertEquals("", details.validateGiverRecipientVisibility(feedbackQuestionAttributes));
+
+        ______TS("failure: giver type is not STUDENT");
+        feedbackQuestionAttributes.setGiverType(FeedbackParticipantType.SELF);
+        assertEquals(FeedbackContributionQuestionDetails.CONTRIB_ERROR_INVALID_FEEDBACK_PATH,
+                details.validateGiverRecipientVisibility(feedbackQuestionAttributes));
+
+        ______TS("failure: recipient type can only be OWN_TEAM_MEMBERS_INCLUDING_SELF");
+        feedbackQuestionAttributes.setGiverType(FeedbackParticipantType.STUDENTS);
+        feedbackQuestionAttributes.setRecipientType(FeedbackParticipantType.SELF);
+        assertEquals(FeedbackContributionQuestionDetails.CONTRIB_ERROR_INVALID_FEEDBACK_PATH,
+                details.validateGiverRecipientVisibility(feedbackQuestionAttributes));
+
+        ______TS("failure: giver type is not STUDENT and recipient type is not OWN_TEAM_MEMBERS_INCLUDING_SELF");
+        feedbackQuestionAttributes.setGiverType(FeedbackParticipantType.SELF);
+        feedbackQuestionAttributes.setRecipientType(FeedbackParticipantType.SELF);
+        assertEquals(FeedbackContributionQuestionDetails.CONTRIB_ERROR_INVALID_FEEDBACK_PATH,
+                details.validateGiverRecipientVisibility(feedbackQuestionAttributes));
```

Comment:
do you also wish to test the fallback `giverType` and `recipientType`? It seems to be part of the logic too.

Reply:
Thank you for your advice but I do not understand what does it mean by "fallback `giverType` and `recipientType`"? Can you explain further for me? Thank you!

Reply:
No problem! As you see in the method, the `giverType` is set to `STUDENTS` and `recipientType` is set to `OWN_TEAM_MEMBERS_INCLUDING_SELF` respectively upon detecting error. 
Do you intend to test this too?

Reply:
OK sure! Thank you for the explaination.